packages/vm/lib/evm/opcodes/functions.ts

```diff
@@ -747,6 +747,23 @@ export const handlers: Map<number, OpHandler> = new Map([
       accessStorageEIP2929(runState, keyBuf, true)
       updateSstoreGasEIP1283(runState, found, setLengthLeftStorage(value))
       updateSstoreGasEIP2200(runState, found, setLengthLeftStorage(value), keyBuf)
+
+      // if we are before Constantinople, then neither SSTORE charges of above have been ran: we charge them here.
+      if (!runState._common.gteHardfork('constantinople')) {
+        const sstoreResetCost = runState._common.param('gasPrices', 'sstoreReset')
+        if ((value.length === 0 && !found.length) || (value.length !== 0 && found.length)) {
+          runState.eei.useGas(
+            new BN(adjustSstoreGasEIP2929(runState, keyBuf, sstoreResetCost, 'reset'))
+          )
+        } else if (value.length === 0 && found.length) {
+          runState.eei.useGas(
+            new BN(adjustSstoreGasEIP2929(runState, keyBuf, sstoreResetCost, 'reset'))
+          )
+          runState.eei.refundGas(new BN(runState._common.param('gasPrices', 'sstoreRefund')))
+        } else if (value.length !== 0 && !found.length) {
+          runState.eei.useGas(new BN(runState._common.param('gasPrices', 'sstoreSet')))
+        }
+      }
       await runState.eei.storageStore(keyBuf, value)
```

Comment:
Ah yes this make sense, sorry and thanks! 🙂 

Perhaps this misplaced stuff could also go in a function in `opcodes/utils` as `updateSStoreGas`?
```js
      updateSstoreGas(runState, found, setLengthLeftStorage(value))
      updateSstoreGasEIP1283(runState, found, setLengthLeftStorage(value))
      updateSstoreGasEIP2200(runState, found, setLengthLeftStorage(value), keyBuf)
```

and be gated by 
```js
// Before constantinople
if (!runState._common.gteHardfork('constantinople')){
  ... 
}
```

Reply:
Yep great suggestion! I will also move the HF logic inside the SSTORE, because otherwise it seems that it updates SSTORE gas 3 times (which is not the case)

Reply:
Oh nice, yes that's a lot clearer now 💯